### PR TITLE
EA-51 "Extend EncounterTransaction to include conditions during save"

### PR DESCRIPTION
**Pull request by tomaszmueller**

### CODE REVIEW

**Comment by tomaszmueller** [comment]
@rkorytkowski https://issues.openmrs.org/browse/EA-51

**Comment by djazayeri** [comment]
I presume that someone from Bahmni should review and merge this. @shruthidipali do you want to do this, or get someone else to?

File: api/pom.xml [hunk 1]
```diff
@@ -36,6 +36,22 @@
...
+            <type>jar</type>
+        </dependency>
+
+        <dependency>
+            <groupId>org.openmrs.module</groupId>
+            <artifactId>emrapi-condition-list</artifactId>
+            <version>1.7-SNAPSHOT</version>
+            <scope>test</scope>
+            <type>test-jar</type>
+        </dependency>
```
Comment (rkorytkowski) [inline api/pom.xml]: Please replace those 2 with just one:
```xml
<dependency>
            <groupId>org.openmrs.module</groupId>
            <artifactId>emrapi-condition-list</artifactId>
            <version>${project.parent.version}</version>
            <scope>compile</scope>
            <type>jar</type>
        </dependency>
```

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java [hunk 2]
```diff
@@ -74,7 +78,8 @@ public EmrEncounterServiceImpl(PatientService patientService, VisitService visit
                                    EncounterDispositionServiceHelper encounterDispositionServiceHelper,
                                    EncounterTransactionMapper encounterTransactionMapper,
                                    EncounterProviderServiceHelper encounterProviderServiceHelper,
-                                   @Qualifier(value = "emrOrderService") EmrOrderService emrOrderService) {
+                                   @Qualifier(value = "emrOrderService") EmrOrderService emrOrderService,
+                                   @Qualifier(value = "conditionService") ConditionService conditionService) {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: Since dependencies are declared in moduleApplicationContext.xml, please remove all @Qualifiers from here.

File: condition-list/src/main/java/org/openmrs/api/impl/ConditionServiceImpl.java [hunk 3]
```diff
@@ -89,6 +90,27 @@ public Condition voidCondition(Condition condition, String voidReason) {
     }
 
     @Override
+    public Condition editCondition(Condition condition) {
+        if(conditionDAO.getConditionByUuid(condition.getUuid()) == null)
+        {
+            throw new IllegalArgumentException("condition doesn't exist");
+        };
+        return conditionDAO.saveOrUpdate(condition);
```
Comment (rkorytkowski) [inline condition-list/src/main/java/org/openmrs/api/impl/ConditionServiceImpl.java]: There's save method already. Why did you add this one?

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java [hunk 4]
```diff
@@ -276,4 +285,53 @@ private Visit findOrCreateVisit(EncounterTransaction encounterTransaction, Patie
...
+        return conditionService.getConditionByUuid(Uuid);
+    }
+
+    private Condition voidCondition(EncounterTransaction encounterTransaction, Condition condition, String voidReason) {
+        return conditionService.voidCondition(condition,voidReason);
+    }
+
+    private void endCondition(Condition condition, Date endDate, Concept endReason) {
+        conditionService.endCondition(condition, endDate, endReason);
+    }
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: I assume you are still working on those methods... They are not used anywhere...

**Comment by tomaszmueller** [comment]
@rkorytkowski updated

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java (hunk 4 above)
Comment (tomaszmueller) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: @rkorytkowski They are not used currently, but I assumed that methods for add/edit/revise/void/end a condition should be present according to the issue description.

Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: The description says "Should internally use the conditionService APIs to add/edit/revise/void/end a condition." meaning that you should call methods in ConditionService whenever relevant and not create them in EmrEncounterServiceImpl.

**Comment by rkorytkowski** [comment]
I'll step away from reviewing this further. I'm not familiar enough with EncounterTransaction to say if Tomasz added condition lists the way it was intended.

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java [hunk 5]
```diff
@@ -276,4 +285,53 @@ private Visit findOrCreateVisit(EncounterTransaction encounterTransaction, Patie
...
+        conditionService.save(condition);
+    }
+
+    private void replaceCondition(List<Condition> conditionList, Condition conditionToReplace, Condition condition) {
+        if(conditionToReplace != null && conditionList.contains(conditionToReplace)) {
+            Collections.replaceAll(conditionList, conditionToReplace, condition);
+        }
+    }
+
+    public Condition reviseCondition(String Uuid) {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: It doesn't make sense to introduce methods that are in ConditionService already. Delete it.

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java [hunk 6]
```diff
@@ -276,4 +285,53 @@ private Visit findOrCreateVisit(EncounterTransaction encounterTransaction, Patie
...
+        return conditionService.getConditionByUuid(Uuid);
+    }
+
+    public Condition voidCondition(EncounterTransaction encounterTransaction, Condition condition, String voidReason) {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: To remove.

File: api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java [hunk 7]
```diff
@@ -276,4 +285,53 @@ private Visit findOrCreateVisit(EncounterTransaction encounterTransaction, Patie
...
+        return conditionService.voidCondition(condition,voidReason);
+    }
+
+    public void endCondition(Condition condition, Date endDate, Concept endReason) {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/emrapi/encounter/EmrEncounterServiceImpl.java]: Remove.

**Comment by tomaszmueller** [comment]
updated.
@djazayeri , @shruthidipali I am waiting for a further review on this issue.

**Comment by dkayiwa** [comment]
@shruthidipali are you still interested in this? Or should we just close it? :)

**Comment by dkayiwa** [comment]
Closing this because it was abandoned.
